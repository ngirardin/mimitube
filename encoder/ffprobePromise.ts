import Ffmpeg, { FfprobeData } from "fluent-ffmpeg";

const probe = async (file: string): Promise<FfprobeData> => {
  return new Promise((resolve, reject) => {
    Ffmpeg.ffprobe(file, (err: any, data: any) => {
      if (err) {
        return reject(err);
      }

      return resolve(data);
    });
  });
};

const getCreationTime = async (metadata: FfprobeData): Promise<Date> => {
  const tags: any = metadata.format.tags;
  const creationTime: string = tags.creation_time;

  if (creationTime.length !== 27) {
    throw new Error("No creation_time tag");
  }

  return new Date(creationTime);
};

export default { probe, getCreationTime };
