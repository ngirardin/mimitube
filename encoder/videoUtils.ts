import fs from "fs/promises";
import ffprobeUtils from "./ffprobeUtils";

interface CreationTime {
  source: "metadata" | "fileCreation";
  date: Date;
}

export default {
  getCreationTime: async (file: string): Promise<CreationTime> => {
    const metadata = await ffprobeUtils.probe(file);
    const creationTime = await ffprobeUtils.getCreationTime(metadata);

    if (creationTime) {
      return { source: "metadata", date: creationTime };
    }

    const { mtime } = await fs.stat(file);

    return { source: "fileCreation", date: mtime };
  },
};
