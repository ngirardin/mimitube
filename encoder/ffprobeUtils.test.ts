import ffprobePromise from "./ffprobeUtils";
import { testVideos } from "./videoUtils.test";

describe("the getCreationTime method", () => {
  it("should return undefine for a Mavic Air video", async () => {
    const probeData = await ffprobePromise.probe(testVideos.drone);
    expect(probeData).toMatchSnapshot();
    expect(await ffprobePromise.getCreationTime(probeData)).toBeUndefined();
  });

  it("should extract the creation date from a gopro video", async () => {
    const probeData = await ffprobePromise.probe(testVideos.gopro);
    expect(probeData).toMatchSnapshot();
    expect(await ffprobePromise.getCreationTime(probeData)).toEqual(new Date("2020-09-13T16:00:25.000Z"));
  });

  it("should extract the creation date from a gsm video", async () => {
    const probeData = await ffprobePromise.probe(testVideos.gsm);
    expect(probeData).toMatchSnapshot();
    expect(await ffprobePromise.getCreationTime(probeData)).toEqual(new Date("2020-11-07T12:09:04.000Z"));
  });
});
