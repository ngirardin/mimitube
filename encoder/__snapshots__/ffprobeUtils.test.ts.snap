// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the getCreationTime method should extract the creation date from a gopro video 1`] = `
Object {
  "chapters": Array [],
  "format": Object {
    "bit_rate": 60392564,
    "duration": 26.837333,
    "filename": "/mnt/c/Users/ngira/Desktop/gopro/GX013175.MP4",
    "format_long_name": "QuickTime / MOV",
    "format_name": "mov,mp4,m4a,3gp,3g2,mj2",
    "nb_programs": 0,
    "nb_streams": 5,
    "probe_score": 100,
    "size": 202596921,
    "start_time": 0,
    "tags": Object {
      "compatible_brands": "mp41",
      "creation_time": "2020-09-13T16:00:25.000000Z",
      "firmware": "HD7.01.01.90.00",
      "major_brand": "mp41",
      "minor_version": "538120216",
    },
  },
  "streams": Array [
    Object {
      "avg_frame_rate": "50/1",
      "bit_rate": 60176876,
      "bits_per_raw_sample": "N/A",
      "chroma_location": "unspecified",
      "codec_long_name": "H.265 / HEVC (High Efficiency Video Coding)",
      "codec_name": "hevc",
      "codec_tag": "0x31637668",
      "codec_tag_string": "hvc1",
      "codec_time_base": "1/50",
      "codec_type": "video",
      "coded_height": 2160,
      "coded_width": 3840,
      "color_primaries": "bt709",
      "color_range": "pc",
      "color_space": "bt709",
      "color_transfer": "bt709",
      "display_aspect_ratio": "16:9",
      "disposition": Object {
        "attached_pic": 0,
        "clean_effects": 0,
        "comment": 0,
        "default": 1,
        "dub": 0,
        "forced": 0,
        "hearing_impaired": 0,
        "karaoke": 0,
        "lyrics": 0,
        "original": 0,
        "timed_thumbnails": 0,
        "visual_impaired": 0,
      },
      "duration": 26.82,
      "duration_ts": 2413800,
      "field_order": "unknown",
      "has_b_frames": 0,
      "height": 2160,
      "id": "N/A",
      "index": 0,
      "level": 153,
      "max_bit_rate": "N/A",
      "nb_frames": 1341,
      "nb_read_frames": "N/A",
      "nb_read_packets": "N/A",
      "pix_fmt": "yuvj420p",
      "profile": "Main",
      "r_frame_rate": "50/1",
      "refs": 1,
      "sample_aspect_ratio": "1:1",
      "start_pts": 0,
      "start_time": 0,
      "tags": Object {
        "creation_time": "2020-09-13T16:00:25.000000Z",
        "encoder": "GoPro H.265 encoder",
        "handler_name": "GoPro H.265",
        "language": "eng",
        "timecode": "16:00:25:04",
      },
      "time_base": "1/90000",
      "timecode": "N/A",
      "width": 3840,
    },
    Object {
      "avg_frame_rate": "0/0",
      "bit_rate": 189250,
      "bits_per_raw_sample": "N/A",
      "bits_per_sample": 0,
      "channel_layout": "stereo",
      "channels": 2,
      "codec_long_name": "AAC (Advanced Audio Coding)",
      "codec_name": "aac",
      "codec_tag": "0x6134706d",
      "codec_tag_string": "mp4a",
      "codec_time_base": "1/48000",
      "codec_type": "audio",
      "disposition": Object {
        "attached_pic": 0,
        "clean_effects": 0,
        "comment": 0,
        "default": 1,
        "dub": 0,
        "forced": 0,
        "hearing_impaired": 0,
        "karaoke": 0,
        "lyrics": 0,
        "original": 0,
        "timed_thumbnails": 0,
        "visual_impaired": 0,
      },
      "duration": 26.837333,
      "duration_ts": 1288192,
      "id": "N/A",
      "index": 1,
      "max_bit_rate": 48000,
      "nb_frames": 1258,
      "nb_read_frames": "N/A",
      "nb_read_packets": "N/A",
      "profile": "LC",
      "r_frame_rate": "0/0",
      "sample_fmt": "fltp",
      "sample_rate": 48000,
      "start_pts": 0,
      "start_time": 0,
      "tags": Object {
        "creation_time": "2020-09-13T16:00:25.000000Z",
        "handler_name": "GoPro AAC  ",
        "language": "eng",
        "timecode": "16:00:25:04",
      },
      "time_base": "1/48000",
    },
    Object {
      "avg_frame_rate": "50/1",
      "bit_rate": 1,
      "bits_per_raw_sample": "N/A",
      "codec_long_name": "unknown",
      "codec_name": "unknown",
      "codec_tag": "0x64636d74",
      "codec_tag_string": "tmcd",
      "codec_type": "data",
      "disposition": Object {
        "attached_pic": 0,
        "clean_effects": 0,
        "comment": 0,
        "default": 1,
        "dub": 0,
        "forced": 0,
        "hearing_impaired": 0,
        "karaoke": 0,
        "lyrics": 0,
        "original": 0,
        "timed_thumbnails": 0,
        "visual_impaired": 0,
      },
      "duration": 26.82,
      "duration_ts": 2413800,
      "id": "N/A",
      "index": 2,
      "max_bit_rate": "N/A",
      "nb_frames": 1,
      "nb_read_frames": "N/A",
      "nb_read_packets": "N/A",
      "profile": "unknown",
      "r_frame_rate": "0/0",
      "start_pts": 0,
      "start_time": 0,
      "tags": Object {
        "creation_time": "2020-09-13T16:00:25.000000Z",
        "handler_name": "GoPro TCD  ",
        "language": "eng",
        "timecode": "16:00:25:04",
      },
      "time_base": "1/90000",
    },
    Object {
      "avg_frame_rate": "0/0",
      "bit_rate": 32364,
      "bits_per_raw_sample": "N/A",
      "codec_long_name": "binary data",
      "codec_name": "bin_data",
      "codec_tag": "0x646d7067",
      "codec_tag_string": "gpmd",
      "codec_type": "data",
      "disposition": Object {
        "attached_pic": 0,
        "clean_effects": 0,
        "comment": 0,
        "default": 1,
        "dub": 0,
        "forced": 0,
        "hearing_impaired": 0,
        "karaoke": 0,
        "lyrics": 0,
        "original": 0,
        "timed_thumbnails": 0,
        "visual_impaired": 0,
      },
      "duration": 26.837,
      "duration_ts": 26837,
      "id": "N/A",
      "index": 3,
      "max_bit_rate": "N/A",
      "nb_frames": 26,
      "nb_read_frames": "N/A",
      "nb_read_packets": "N/A",
      "profile": "unknown",
      "r_frame_rate": "0/0",
      "start_pts": 0,
      "start_time": 0,
      "tags": Object {
        "creation_time": "2020-09-13T16:00:25.000000Z",
        "handler_name": "GoPro MET  ",
        "language": "eng",
      },
      "time_base": "1/1000",
    },
    Object {
      "avg_frame_rate": "0/0",
      "bit_rate": 12765,
      "bits_per_raw_sample": "N/A",
      "codec_long_name": "unknown",
      "codec_name": "unknown",
      "codec_tag": "0x63736466",
      "codec_tag_string": "fdsc",
      "codec_type": "data",
      "disposition": Object {
        "attached_pic": 0,
        "clean_effects": 0,
        "comment": 0,
        "default": 1,
        "dub": 0,
        "forced": 0,
        "hearing_impaired": 0,
        "karaoke": 0,
        "lyrics": 0,
        "original": 0,
        "timed_thumbnails": 0,
        "visual_impaired": 0,
      },
      "duration": 26.82,
      "duration_ts": 2413800,
      "id": "N/A",
      "index": 4,
      "max_bit_rate": "N/A",
      "nb_frames": 2627,
      "nb_read_frames": "N/A",
      "nb_read_packets": "N/A",
      "profile": "unknown",
      "r_frame_rate": "0/0",
      "start_pts": 0,
      "start_time": 0,
      "tags": Object {
        "creation_time": "2020-09-13T16:00:25.000000Z",
        "handler_name": "GoPro SOS  ",
        "language": "eng",
      },
      "time_base": "1/90000",
    },
  ],
}
`;

exports[`the getCreationTime method should extract the creation date from a gsm video 1`] = `
Object {
  "chapters": Array [],
  "format": Object {
    "bit_rate": 38838036,
    "duration": 40.434,
    "filename": "/mnt/c/Users/ngira/Desktop/gopro/VID_20201107_130823.mp4",
    "format_long_name": "QuickTime / MOV",
    "format_name": "mov,mp4,m4a,3gp,3g2,mj2",
    "nb_programs": 0,
    "nb_streams": 2,
    "probe_score": 100,
    "size": 196297144,
    "start_time": 0,
    "tags": Object {
      "com.android.capture.fps": "60.000000",
      "com.android.version": "10",
      "compatible_brands": "isommp42",
      "creation_time": "2020-11-07T12:09:04.000000Z",
      "location": "+48.5687+007.7474/",
      "location-eng": "+48.5687+007.7474/",
      "major_brand": "mp42",
      "minor_version": "0",
    },
  },
  "streams": Array [
    Object {
      "avg_frame_rate": "218340000/3638761",
      "bit_rate": 38584742,
      "bits_per_raw_sample": "N/A",
      "chroma_location": "unspecified",
      "codec_long_name": "H.265 / HEVC (High Efficiency Video Coding)",
      "codec_name": "hevc",
      "codec_tag": "0x31637668",
      "codec_tag_string": "hvc1",
      "codec_time_base": "3638761/218340000",
      "codec_type": "video",
      "coded_height": 2176,
      "coded_width": 3840,
      "color_primaries": "bt470bg",
      "color_range": "pc",
      "color_space": "bt470bg",
      "color_transfer": "smpte170m",
      "display_aspect_ratio": "16:9",
      "disposition": Object {
        "attached_pic": 0,
        "clean_effects": 0,
        "comment": 0,
        "default": 1,
        "dub": 0,
        "forced": 0,
        "hearing_impaired": 0,
        "karaoke": 0,
        "lyrics": 0,
        "original": 0,
        "timed_thumbnails": 0,
        "visual_impaired": 0,
      },
      "duration": 40.430678,
      "duration_ts": 3638761,
      "field_order": "unknown",
      "has_b_frames": 0,
      "height": 2160,
      "id": "N/A",
      "index": 0,
      "level": 153,
      "max_bit_rate": "N/A",
      "nb_frames": 2426,
      "nb_read_frames": "N/A",
      "nb_read_packets": "N/A",
      "pix_fmt": "yuvj420p",
      "profile": "Main",
      "r_frame_rate": "60/1",
      "refs": 1,
      "sample_aspect_ratio": "1:1",
      "start_pts": 0,
      "start_time": 0,
      "tags": Object {
        "creation_time": "2020-11-07T12:09:04.000000Z",
        "handler_name": "VideoHandle",
        "language": "eng",
      },
      "time_base": "1/90000",
      "timecode": "N/A",
      "width": 3840,
    },
    Object {
      "avg_frame_rate": "0/0",
      "bit_rate": 96118,
      "bits_per_raw_sample": "N/A",
      "bits_per_sample": 0,
      "channel_layout": "stereo",
      "channels": 2,
      "codec_long_name": "AAC (Advanced Audio Coding)",
      "codec_name": "aac",
      "codec_tag": "0x6134706d",
      "codec_tag_string": "mp4a",
      "codec_time_base": "1/48000",
      "codec_type": "audio",
      "disposition": Object {
        "attached_pic": 0,
        "clean_effects": 0,
        "comment": 0,
        "default": 1,
        "dub": 0,
        "forced": 0,
        "hearing_impaired": 0,
        "karaoke": 0,
        "lyrics": 0,
        "original": 0,
        "timed_thumbnails": 0,
        "visual_impaired": 0,
      },
      "duration": 40.431333,
      "duration_ts": 1940704,
      "id": "N/A",
      "index": 1,
      "max_bit_rate": 96000,
      "nb_frames": 1895,
      "nb_read_frames": "N/A",
      "nb_read_packets": "N/A",
      "profile": "LC",
      "r_frame_rate": "0/0",
      "sample_fmt": "fltp",
      "sample_rate": 48000,
      "start_pts": 0,
      "start_time": 0,
      "tags": Object {
        "creation_time": "2020-11-07T12:09:04.000000Z",
        "handler_name": "SoundHandle",
        "language": "eng",
      },
      "time_base": "1/48000",
    },
  ],
}
`;

exports[`the getCreationTime method should return undefine for a Mavic Air video 1`] = `
Object {
  "chapters": Array [],
  "format": Object {
    "bit_rate": 103346150,
    "duration": 290.991,
    "filename": "/mnt/c/Users/ngira/Desktop/gopro/drone/DJI_0009.MP4",
    "format_long_name": "QuickTime / MOV",
    "format_name": "mov,mp4,m4a,3gp,3g2,mj2",
    "nb_programs": 0,
    "nb_streams": 1,
    "probe_score": 100,
    "size": 3759099952,
    "start_time": 0,
    "tags": Object {
      "compatible_brands": "isomiso2avc1mp41",
      "encoder": "Lavf56.15.102",
      "major_brand": "isom",
      "minor_version": "512",
    },
  },
  "streams": Array [
    Object {
      "avg_frame_rate": "30000/1001",
      "bit_rate": 103344564,
      "bits_per_raw_sample": 8,
      "chroma_location": "left",
      "codec_long_name": "H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10",
      "codec_name": "h264",
      "codec_tag": "0x31637661",
      "codec_tag_string": "avc1",
      "codec_time_base": "1001/60000",
      "codec_type": "video",
      "coded_height": 2160,
      "coded_width": 3840,
      "color_primaries": "unknown",
      "color_range": "unknown",
      "color_space": "unknown",
      "color_transfer": "unknown",
      "display_aspect_ratio": "N/A",
      "disposition": Object {
        "attached_pic": 0,
        "clean_effects": 0,
        "comment": 0,
        "default": 1,
        "dub": 0,
        "forced": 0,
        "hearing_impaired": 0,
        "karaoke": 0,
        "lyrics": 0,
        "original": 0,
        "timed_thumbnails": 0,
        "visual_impaired": 0,
      },
      "duration": 290.9907,
      "duration_ts": 8729721,
      "field_order": "unknown",
      "has_b_frames": 0,
      "height": 2160,
      "id": "N/A",
      "index": 0,
      "is_avc": "true",
      "level": 51,
      "max_bit_rate": "N/A",
      "nal_length_size": 4,
      "nb_frames": 8721,
      "nb_read_frames": "N/A",
      "nb_read_packets": "N/A",
      "pix_fmt": "yuv420p",
      "profile": "High",
      "r_frame_rate": "30000/1001",
      "refs": 1,
      "sample_aspect_ratio": "N/A",
      "start_pts": 0,
      "start_time": 0,
      "tags": Object {
        "handler_name": "VideoHandler",
        "language": "und",
      },
      "time_base": "1/30000",
      "timecode": "N/A",
      "width": 3840,
    },
  ],
}
`;
